grails-app/services/com/unifina/service/PermissionService.groovy

```diff
@@ -97,19 +97,42 @@ class PermissionService {
 		}.toList()
 
 		// Special case of UI channels: they inherit permissions from the associated canvas
-		Set<Permission> syntheticPermissions = new HashSet<>()
 		if (resource instanceof Stream && resource.uiChannel) {
+			Set<Permission> syntheticPermissions = new HashSet<>()
+			Permission permission = hasTransitiveDashboardPermissions(resource.uiChannelCanvas, userish)
```

Comment:
There's actually a vulnerability here. Test case:

- Create canvas
- Two modules: public (id 1) and secret (id 2). The modules have ui channel streams: module id 1 has stream `uiChannelPublic` and module 2 has stream `uiChannelSecret`.
- Create Dashboard
- Grant user `U` `DASHBOARD_GET` permission to dashboard
- Create DashboardItem with a reference to module id 1 (public).
- assert `getPermissionsTo(uiChannelPublic, U) should return `STREAM_GET` and `STREAM_SUBSCRIBE`.
- assert `getPermissionsTo(uiChannelSecret, U) should return an empty set

Maybe replace `hasTransitiveDashboardPermissions(Canvas, Userish) with

```
boolean hasPermissionViaDashboard(stream, userish)
```